### PR TITLE
Initial support for mimxrt1020 on evk

### PR DESCRIPTION
**Pull request by mubes**
***For comment, not for pull***

This PR implements TinyUF2 support for MIMXRT1020 series chips, specifically the evk and Versiboard2 in the first instance.

Functionality is identical to the MIMXRT1010 version.

Memory size is left the same as the 1010...there's no advantage in changing it.

### CODE REVIEW

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c]: Should be `* 1000` here too

**Comment by arturo182** [comment]
Looks great, just make sure to fix the build for the 1010 evk (and test if you have one), and add 1020 evk to the build script here https://github.com/arturo182/tinyuf2/blob/master/.github/workflows/build.yml#L11

**Review by arturo182 — CHANGES_REQUESTED** [review]

**Comment by mubes** [comment]
Yeah, I wasn't sure until we got to testing....figured it must've been 
working for you somehow.

I'll incorporate.

DAVE

**Comment by mubes** [comment]
Update. Board configuration is tidied and fully tested on 1020evk.

* Added 'multitap' because in my experience double-tapping can easily be done by accident, and I don't want users back in Kansas without knowing why....four taps _never_ happens by accident :-) on by option setting in the configuration file, you can have up to 255 if you really want to upset your users.
* Inverted the setting of the boot led so you can more easily see when it's in the bootloader. Again, set in the configuration file.
* Made bootmode flashrate configurable..mostly cos I like it quicker, and others will like it slower.

1010 certainly gets into the bootloader and can get back into the bootloader via the double tap or multitap mechanisms. Haven't been able to load a uf2 application yet but don't have a decent debug setup for 1010 at the moment so it's pending.

Issues at the moment are mostly on VB2, which uses the secondary pinmux for QSPI, and so no DQS signalling.

More to be done yet. Still not for pull.

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c:284]: The extra spaces here are by mistake, you can remove them

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c:332]: Please don't align the second argument like that, there's no point ;)

File: hw/bsp/mimxrt1010_evk/board_config.h
Comment (arturo182) [inline hw/bsp/mimxrt1010_evk/board_config.h]: Think the comment is a bit redundant, we know what file we are in :)

**Review by arturo182 — CHANGES_REQUESTED** [review]
I think the board_config.h files are starting to get a bit out of hand, I want them to be as small as possible to it's really easy to add new boards (see https://github.com/adafruit/uf2-samdx1/blob/master/boards/serpente/board_config.h, although I know if we want to support multiple chip families it can't be that simple). 

For QSPI stuff I don't think each board has to specify all that pin data because there's only so many configs for each chip that can exist, so we can have the pin options defined in the chip files and then in board just specify which qspi port to use.

For UART cfg I think it's always gonna be the same regardless of the pin used so that could be moved out of board config as well.

Similar for LED except for the pull direction but that can be deduced from LED_STATE_ON value (pull the opposite way).

For the tap and LED behaviour, I'd actually vote against being able to customize that because we want the same UX on each board that will have the bootloader. I also want to keep the UX as close to the reference uf2 bootloader (https://github.com/adafruit/uf2-samdx1) as possible.

File: Makefile
Comment (arturo182) [inline Makefile:26]: `CHIP_FAMILY` is used by tinyUSB, is there any benefit from renaming it?

File: README.md
Comment (arturo182) [inline README.md]: We can add a "(courtesy of Adafruit Industries)" here :)

File: hw/bsp/bsp.h
Comment (arturo182) [inline hw/bsp/bsp.h:35]: In the samd port these options are in the uf2.h file so I'm a bit torn on where they should go, but I guess seeing as they are board-specific, we might as well have them here

File: hw/bsp/mimxrt1010_evk/board_config.h
Comment (arturo182) [inline hw/bsp/mimxrt1010_evk/board_config.h:9]: Not sure this comment adds much

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c:399]: Took me a moment to understand how we end up here, maybe add a comment about this condition

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c]: "to to"

File: hw/chip/mimxrt10xx/family.c
Comment (arturo182) [inline hw/chip/mimxrt10xx/family.c:460]: So we reset twice, I guess that's ok.

File: hw/bsp/mimxrt1010_evk/board.mk
Comment (arturo182) [inline hw/bsp/mimxrt1010_evk/board.mk:2]: I'm not sure there's any benefit from having this one extra level `TUF2_CHIP_MEMBER` of specification for chip?

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (arturo182) [inline hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h]: Don't think this is the right way to go, think how many different UARTs there are and how many pin configs each can have, depending on the number of boards we could have double-digit amount of defines here. 

The mux and cfg values will always be the same, so we only need the port and PIN_UART_RX and PIN_UART_TX, these can be moved back to the board_config.h

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (arturo182) [inline hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h]: Same here, the GPIO can be on basically any pin, so we would need as many defines as there are pins, only PIN_LED and LED_STATE_ON should be needed, others either don't change or can be inferred from that info (like pull should be the opposite of STATE_ON).

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (arturo182) [inline hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h:65]: This should stay here cause there are only 2 options, although the _CFG will always be same so could go directly into the .c file.

File: hw/chip/mimxrt10xx/mimxrt102x/pinning_options.h
Comment (arturo182) [inline hw/chip/mimxrt10xx/mimxrt102x/pinning_options.h:67]: Same comments as above

**Review by arturo182 — CHANGES_REQUESTED** [review]
Hope my review is not too harsh 😅

File: Makefile
Comment (mubes) [reply Makefile:26]: Only because the other two elements have the TUF2 on the front, showing they're 'owned' by this code...if we adopt a 'style' to make it easier for the reader, then we shouldn't break it on the very same line :-)  CHIP_FAMILY is something different and receives the value 'transdimension' which means there had to be an absolute path in family.mk otherwise it picks up the wrong value. This approach avoids that and allows things to be more standard.

File: README.md
Comment (mubes) [reply README.md]: Agreed, done.

File: hw/bsp/mimxrt1010_evk/board_config.h
Comment (mubes) [reply hw/bsp/mimxrt1010_evk/board_config.h:9]: Because the board may (and probably will, in the general case) have a different VID:PID in run mode....this comment is to make that explicit.

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c:399]: Agreed, added.

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (mubes) [reply hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h]: You were the one wanting to clean up board_config.h :-)  
OK, moved them back and left the MUX and CFG values here, ifdef guarded in case someone wants to override them.

File: Makefile
Comment (arturo182) [reply Makefile:26]: Ok, I'm convinced :)

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (mubes) [reply hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h]: Agreed, same discussion as for the LPUART, with the same resolution. I personally don't like the idea of inferring STATE_ON because it's just complexity ... easier and more transparent just to make it explicit. The next maintainer  might not be as smart as us :-)

File: hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h
Comment (mubes) [reply hw/chip/mimxrt10xx/mimxrt101x/pinning_options.h:65]: Agreed, but lets not move the CFG into the c as magic numbers ... I would prefer to keep the whole CFG/MUX/Settings in one place rather than sending people scurrying through multiple files if they ever need changing (shades of MISRA rules)

File: hw/bsp/mimxrt1010_evk/board.mk
Comment (mubes) [reply hw/bsp/mimxrt1010_evk/board.mk:2]: Unfortunately there is because NXP, bless 'em, decided to pull an ST and make very similar numbered chips subtly different (specifically in pin numbering etc.) The extra level means we can keep the pinning_options, ld and mk files in separate directories per chip and the level above contain everything that is common across the entire family. This will become more important for ST, I fear.

File: hw/bsp/bsp.h
Comment (mubes) [reply hw/bsp/bsp.h:35]: I thought about this a lot and took the view that they're potentially board specific, so belong here.  Consider the case that someone wants to use tinyuf2 but have differentiated naming for _their_ board - this allows them to do that.  Yes, I know that breaks universal look and feel, but folks _will_ want change, and if that change isn't supported by the codebase (even through gritted teeth) then they will fork, and that would be worse.

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c:284]: Fixed.

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c:332]: Fixed.

**Comment by arturo182** [comment]
Also please remember to add your boards to https://github.com/arturo182/tinyuf2/blob/master/.github/workflows/build.yml#L11

**Comment by mubes** [comment]
Fixes applied, take a look at the comment about the three level path and see what you think (especially in the light of the comment about ST chips). Patch won't come until tomorrow because although everything builds clean I want to test on physical h/w before pushing a pullable patch.

File: hw/bsp/mimxrt1010_evk/board.mk
Comment (arturo182) [reply hw/bsp/mimxrt1010_evk/board.mk:2]: Sounds good, let's keep it then.

**Comment by arturo182** [comment]
Sounds great, after we merge I'll add my Feathers too and will see if any fixes are needed.

**Comment by mubes** [comment]
Yeah, adding Versiboard2 helped smoke out a few issues, I'm sure there will be more.

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c]: fixed

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c:460]: Yep, added a note to that effect. Means the platform comes up 'clean' without us having messed with it.

File: hw/chip/mimxrt10xx/family.c
Comment (mubes) [reply hw/chip/mimxrt10xx/family.c]: Grrr. Hidden comments.  Resolved.

**Comment by arturo182** [comment]
All great, except I, would remove the release images, they will get stale as soon as the next commit happens and having binary blobs in git is icky ;)

I think for the releases we will use the Github Releases process and generate a .bin file for each board when we do that, similar to how CPY does this.

**Review by arturo182 — APPROVED** [review]

**Pull request merged**
